### PR TITLE
[networks] Add prototype for HTTP monitoring in eBPF 

### PR DESCRIPTION
**Pull request by p-lambert**
#### Overview

This PR provides a HTTP monitoring prototype using eBPF. In a nutshell, we track HTTP transactions by attaching an eBPF `BPF_PROG_TYPE_SOCKET_FILTER` program to a “headless” `SOCK_RAW` and then inspecting traffic hitting ports 80 and 8080.

From each ethernet frame matching this criteria we extract a `conn_tuple_t` similarly to what we do on the `kprobe` programs.

Then we do a map lookup in order to retrieve the associated `http_transaction_t`. This embeds information such as the state of each transaction (eg. whether we’re processing a request or response), the associated HTTP method, the status code of the response, the latency of the whole transaction as well as a fragment of the request. 

The state machine governing the processing is quite simple:
```
HTTP_UNKNOWN -> HTTP_REQUESTING -> HTTP_RESPONDING
``` 

When there is no state associated with the current TCP connection (`HTTP_UNKNOWN`) we read up to `HTTP_BUFFER_SIZE` bytes from the TCP segment.

If we match `^(GET|POST|PUT|DELETE|HEAD)` we copy the request fragment to `http_transaction`  along with the time of the request and transition to the `HTTP_REQUESTING` state.

For each following segment we inspect again the first `HTTP_BUFFER_SIZE` bytes of the payload until we match `^HTTP` (note that an HTTP request can be split across multiple segments). Once that happens we parse the status code and transition to `HTTP_RESPONDING`.

For each subsequent segment we update the `response_last_seen` timestamp field from the `http_transaction_t` until one of the two conditions happen:
* A TCP segment with a `FIN` flag is detected;
* A new request is detected, that is, we see another TCP segment beginning with `^(GET|POST|PUT|DELETE|HEAD)`  (which can happen in the context of HTTP keep-alives);

Once that happens we consider this HTTP transaction “done” and enqueue the `http_transaction_t` struct in another eBPF map where transactions are batched for consumption from userspace. 

#### Batching

**UPDATE** I'm trying a different approach in https://github.com/DataDog/datadog-agent/pull/6740/files

From our past experience with TCP closed connections (which are flushed using perf events), we know that we need some sort of batching to reduce the number of times we poll the perf ring in order to keep the overhead of system-probe low in high throughput workloads.

Here we take a similar approach and using a `http_batch_t` struct that looks like the following:

batch

Basically each `http_batch_t` entry is a circular buffer divided in multiple “pages”. Once we fill up a whole page with `http_transaction_t` structs, we send a notification `http_notification_t` perf event to the userspace program, which essentially conveys that “page X from batch from CPU Y” can be read. 

There are obviously issues with this approach and I’ll try to explain some of the tradeoffs involved in the following sections. 

Finally there are two params that can be configured regarding the batching mechanism:
* `HTTP_BATCH_SIZE` : determines the size of each “page”. The larger this number is the lower the CPU overhead of consuming the perf events. The downside is increased latency.
* `HTTP_BATCH_PAGES`: determines the number of pages for each batch object. In theory the greater this number is, the less likely is missing events due to a data race. (Having more pages gives more time to the userspace program to read the page before it gets overridden);

#### Why not use TC classifiers?

I haven’t fully thought about this yet, but the main reason we’re not using a TC program here is because you need to attach it to a specific device, and that would be tricky in the context of ephemeral containers.
However one major downside of socket filter programs compared to TC classifiers is that you don’t have access to the `bpf_perf_event_output` helper until [Linux 5.4](https://github.com/torvalds/linux/commit/7c4b90d79d0f4ee6f2c01a114ef0a83a09dfc900). Because of that we have to resort to a hack of flushing the perf events from a `kprobe`. Here we’re using `kretprobe/tcp_sendmsg`  (where `http_notify_batch` is called) but I haven’t thought much about that yet, so there are likely better hook points for that.

In any case most of the code would remain the same regardless of the eBPF program type we decide to use.

#### Why not flush the batch of HTTP transactions directly to the perf ring?

TL;DR to support Kernels <4.11.
The reason why we send a `http_notification_t` object to userspace as opposed to a batch of transactions is because you can’t flush map entries directly to the perf ring until Kernel 4.11. Refer to [this JIRA card](https://datadoghq.atlassian.net/browse/NET-869?focusedCommentId=294037) for the code bits showing this.

Without being able to do that, we would need to copy the batch into the eBPF stack which is limited to 512bytes. I tried to that and we blow up the stack with a batch size of 2.

On the other hand that means that if we drop support for Kernel versions older than 4.11 we can simplify the batching logic by a lot: we would get rid of the map lookup from userspace, the `http_notification_t` bits, and wouldn’t need the multi page stuff as we wouldn’t need a circular buffer. 

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

Write there any instructions and details you may have to test your PR.

### CODE REVIEW

File: pkg/ebpf/c/bpf_helpers.h
Comment (brycekahle) [inline pkg/ebpf/c/bpf_helpers.h:57]: Note: This is only correct for little endian systems. See https://github.com/torvalds/linux/blob/47ec5303d73ea344e84f46660fff693c57641386/tools/lib/bpf/bpf_endian.h#L50-L78

File: pkg/ebpf/c/tracer-ebpf.c
Comment (leeavital) [inline pkg/ebpf/c/tracer-ebpf.c:1320]: I wonder if we could use one of the atomic intrinsics to avoid writing to txs[i] while userspace reads it. 

So userspace could write FREE to txs[i] and the kprobe could do  

```
if ! ( compare_and_swap(txs[i], FREED, WRITING) ) {
    // increment counter for dropped HTTP event
} else {
   __builtin_memcpy(&batch->txs[i], http, sizeof(http_transaction_t));
}
```

File: pkg/ebpf/c/tracer-ebpf.c
Comment (brycekahle) [inline pkg/ebpf/c/tracer-ebpf.c:1365]: If the intention is to only capture the "class" of status code, then we should be careful to label it appropriately. I could see calling it `response_code` being misleading, if it was only 1xx/2xx/3xx/4xx/5xx.

File: pkg/ebpf/c/tracer-ebpf.c
Comment (brycekahle) [inline pkg/ebpf/c/tracer-ebpf.c:1396]: Are we planning to support the other common methods? `CONNECT`, `OPTIONS`, and `PATCH`?

File: pkg/ebpf/c/tracer-ebpf.c
Comment (leeavital) [inline pkg/ebpf/c/tracer-ebpf.c:1358]: Will this pick up error codes like 404 ? It looks like it only takes the first digit into account

File: pkg/ebpf/c/tracer-ebpf.c
Comment (leeavital) [reply pkg/ebpf/c/tracer-ebpf.c:1320]: nevermind, this relies on being able to write to part of the batch from userspace which isn't possible

File: pkg/network/socket_filter.go
Comment (sunhay) [inline pkg/network/socket_filter.go:28]: Besides the point of this PR, but this is a bit awkward  to read - we could probably just change the function argument interface so its written as:

```go
err := util.WithRootNS(rootPath, func() err {
	packetSrc, srcErr = newPacketSource(filter)
	return srcErr
})
if err != nil {
	return nil, err
}
```

where `err = nsErr or srcErr` within `util.WithRootNS()`

File: pkg/ebpf/c/tracer-ebpf.h
Comment (brycekahle) [inline pkg/ebpf/c/tracer-ebpf.h:86]: This feels confusing, as it is mixing two concerns:
1. the state of the transaction: unknown->requesting->responding
2. The request method

It is also referenced in code as both `state` and `packet_type`. Perhaps they should be separate things, or at least different masked parts of the byte.

Method also doesn't need to be a bitmask, since it cannot be multiple methods at the same time.

File: pkg/ebpf/http.go
Comment (brycekahle) [inline pkg/ebpf/http.go:39]: Since it seems you concerned about performance of this function (due to the loops below, rather than something like `strings.Split`), if you aren't aware,  `C.GoBytes` will copy (and thus allocate). If you want to iterate on the raw bytes, you can instead do `(*[C.HTTP_BUFFER_SIZE]byte)(unsafe.Pointer(&tx.request_fragment))` and then only do a `copy` once you've found the range you are interested in.

You may also want to cache that return value with a `sync.Once` or something, if this function may be called a lot.

File: pkg/ebpf/http.go
Comment (brycekahle) [inline pkg/ebpf/http.go:137]: We probably want to rename `ClosedChannel` to something more generic, since Closed doesn't make much sense in this context.

File: pkg/ebpf/http.go
Comment (brycekahle) [inline pkg/ebpf/http.go:171]: Should this include `hits[100]`?

File: pkg/ebpf/c/bpf_helpers.h
Comment (p-lambert) [reply pkg/ebpf/c/bpf_helpers.h:57]: Good catch 👍

File: pkg/ebpf/c/tracer-ebpf.c
Comment (p-lambert) [reply pkg/ebpf/c/tracer-ebpf.c:1365]: Agreed. I think we actually want the real status code, so I'll probably fix that instead of renaming this

File: pkg/network/socket_filter.go
Comment (p-lambert) [reply pkg/network/socket_filter.go:28]: Agreed 👍 I don't mind changing that in this PR

File: pkg/ebpf/c/tracer-ebpf.c
Comment (brycekahle) [reply pkg/ebpf/c/tracer-ebpf.c:1365]: If we want to keep consistent naming with the standard, these are called "status codes".

File: pkg/network/socket_filter.go
Comment (p-lambert) [reply pkg/network/socket_filter.go]: Done

File: pkg/network/ebpf/c/prebuilt/tracer.c
Comment (brycekahle) [inline pkg/network/ebpf/c/prebuilt/tracer.c:693]: Why is this being put into a debug-only block?

File: pkg/network/ebpf/c/prebuilt/tracer.c
Comment (brycekahle) [inline pkg/network/ebpf/c/prebuilt/tracer.c:1269]: You still have a TODO here for IPv6

File: pkg/network/ebpf/c/prebuilt/tracer.c
Comment (brycekahle) [inline pkg/network/ebpf/c/prebuilt/tracer.c:1397]: Did you want to `break` after the `memcpy`?

File: pkg/network/ebpf/c/prebuilt/tracer.c
Comment (brycekahle) [inline pkg/network/ebpf/c/prebuilt/tracer.c:1438]: Did you want to `break` once `space_found && status_code >= 100`?

File: pkg/network/ebpf/c/prebuilt/tracer.c
Comment (brycekahle) [inline pkg/network/ebpf/c/prebuilt/tracer.c:1476]: We are missing the `OPTIONS` and `PATCH` methods. `OPTIONS` is quite common because of CORS requests.

File: pkg/network/ebpf/c/tracer.h
Comment (brycekahle) [inline pkg/network/ebpf/c/tracer.h:90]: I mentioned this before, but I'll post it again:

This feels confusing, as it is mixing two concerns:

the state of the transaction: unknown->requesting->responding
The request method
It is also referenced in code as both state and packet_type. Perhaps they should be separate things, or at least different masked parts of the byte.

Method also doesn't need to be a bitmask, since it cannot be multiple methods at the same time.

File: pkg/network/http/monitor.go
Comment (brycekahle) [inline pkg/network/http/monitor.go:99]: You should probably `defer report.Stop()` to cleanup the `time.Ticker`.

**Review by brycekahle — CHANGES_REQUESTED** [review]

File: pkg/network/ebpf/c/prebuilt/tracer.c
Comment (p-lambert) [reply pkg/network/ebpf/c/prebuilt/tracer.c:1397]: yes, but it doesn't work.
I can't remember now if the problem was with the loop unrolling itself or if the it was something with the verifier

File: pkg/network/ebpf/c/prebuilt/tracer.c
Comment (p-lambert) [reply pkg/network/ebpf/c/prebuilt/tracer.c:1438]: Same as above

File: pkg/network/ebpf/c/prebuilt/tracer.c
Comment (p-lambert) [reply pkg/network/ebpf/c/prebuilt/tracer.c:693]: this `kretprobe` was being used for debug purposes before. now that this is being shared with the http program, I wrapped it in debug block to minimize the overhead when http monitoring is on but debug is off.

File: pkg/network/ebpf/c/prebuilt/tracer.c
Comment (p-lambert) [reply pkg/network/ebpf/c/prebuilt/tracer.c:1489]: Done

File: pkg/network/ebpf/c/tracer.h
Comment (p-lambert) [reply pkg/network/ebpf/c/tracer.h]: Done. Let me know if the code now is a bit more clear

File: pkg/network/ebpf/c/prebuilt/tracer.c
Comment (p-lambert) [reply pkg/network/ebpf/c/prebuilt/tracer.c:1269]: I think it's fine to merge it without IPv6 support for now

Comment (brycekahle) [reply pkg/network/ebpf/c/prebuilt/tracer.c:1269]: This would be a regression because DNS supports IPv6 right now.

Comment (p-lambert) [reply pkg/network/ebpf/c/prebuilt/tracer.c:1269]: Oh, that's a very good point. I forgot this code path is shared 😳

**Review by olivielpeau — APPROVED** [review]
`Datadog/agent-core` seems to be set as a codeowner of all the files in the PR, although it should only own `pkg/config/config.go`, not sure why as the CODEOWNERS file seems to be up-to-date...

In any case, approving for agent-core, only relevant change for agent-core is in `pkg/config/config.go`.

Please make sure the gitlab pipeline passes before merging. The `arm64` system-probe build is now fixed on master so please pick up the latest changes in master in your branch.

File: pkg/network/ebpf/c/prebuilt/tracer.c
Comment (p-lambert) [reply pkg/network/ebpf/c/prebuilt/tracer.c:1269]: I opened a PR with an integration test for DNS on IPv6 to make sure there won't be any regressions: https://github.com/DataDog/datadog-agent/pull/6944

**Review by sunhay — APPROVED** [review]
👍 for the /pkg/process config changes

**Review by brycekahle — APPROVED** [review]

**Pull request merged**
